keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -299,31 +298,39 @@ class ApplicationService(
           environmentSummary.getArtifactPromotionStatus(artifact, artifactVersion.version)
             ?.let { status ->
               if (artifact.isUsedIn(environment)) { // only add a summary if the artifact is used in the environment
+                val artifactInEnvStartTime = now
                 buildArtifactSummaryInEnvironment(
```

Comment:
I find the number of parameters to this function to be unwieldy at this point. It reminded of the verification stuff where we debated this in a PR review and Lorin introduced a `VerificationContext`, which I thought was create. I wonder if we could generalize that class a bit to be used elsewhere and hold these related context bits (delivery config, environment, artifact at least, but possibly other bits as well).

I also wondered if we could simplify this a bit further by using just `artifactVersionInfoInEnvironment` and ditching `artifactVersions`. I wasn't sure why you're filtering for only `CURRENT` and `PREVIOUS` in the query to build the artifact info, but maybe you could consider removing that filter from the query so you could use a single list here?

Reply:
I don't think there's enough benefit to doing the kind of refactoring you're mentioning. The goal of this PR is to change the minimum possible things and see if it improves our API time. I'll likely refactor this whole method in the near future to support the API work that Rani wants. 

Some of this PR (like the `getArtifactVersionByPromotionStatus(...)` functions) have been lifted from the SQL repository to optimize. For this case (filtering by current and previous) there's a clear function comment as to why (`//only CURRENT and PREVIOUS are supported, as they can be sorted by deploy_at`).